service/src/main/java/com/codeforcommunity/processor/TeamsProcessorImpl.java

```diff
@@ -35,4 +74,279 @@ public void disbandTeam(JWTData userData, int teamId) {
+      }
+    } else {
+      throw new WrongTeamRoleException(
+          leaderTeamsRecord.getTeamId(), leaderTeamsRecord.getTeamRole());
+    }
+  }
+
+  @Override
+  public void kickUser(JWTData userData, int teamId, int memberId) {
+    checkUserExists(memberId);
+    checkTeamExists(teamId);
+    UsersTeamsRecord leaderTeamsRecord =
+        db.selectFrom(USERS_TEAMS)
+            .where(USERS_TEAMS.USER_ID.eq(userData.getUserId()))
+            .and(USERS_TEAMS.TEAM_ID.eq(teamId))
+            .fetchOne();
+    if (leaderTeamsRecord != null && leaderTeamsRecord.getTeamRole() == TeamRole.LEADER) {
+      UsersTeamsRecord approvedUserRecord =
+          db.selectFrom(USERS_TEAMS.where(USERS_TEAMS.USER_ID.eq(memberId)))
+              .fetchOne();
+      if (approvedUserRecord.getTeamRole() == TeamRole.MEMBER) {
+        approvedUserRecord.setTeamRole(TeamRole.None);
+        approvedUserRecord.update();
+      } else {
+        throw new MemberStatusException(memberId, teamId);
+      }
+    } else {
+      throw new WrongTeamRoleException(
+          leaderTeamsRecord.getTeamId(), leaderTeamsRecord.getTeamRole());
+    }
+  }
+
+  @Override
+  public void leaveTeam(JWTData userData, int teamId) {
+    checkTeamExists(teamId);
+    UsersTeamsRecord usersTeamsRecord =
+        db.selectFrom(USERS_TEAMS)
+            .where(USERS_TEAMS.USER_ID.eq(userData.getUserId()))
+            .and(USERS_TEAMS.TEAM_ID.eq(teamId))
+            .fetchOne();
+    if (usersTeamsRecord != null && usersTeamsRecord.getTeamRole() != TeamRole.LEADER) {
+      usersTeamsRecord.setTeamRole(TeamRole.None);
+      usersTeamsRecord.update();
+    } else {
+      throw new MemberStatusException(teamId, userData.getUserId());
+    }
+  }
+
+  @Override
+  public void transferOwnership(JWTData userData, TransferOwnershipRequest transferOwnershipRequest, int teamId) {
+    checkUserExists(transferOwnershipRequest.getNewLeaderId());
+    checkTeamExists(teamId);
+    UsersTeamsRecord oldLeaderTeamsRecord =
+        db.selectFrom(USERS_TEAMS)
+            .where(USERS_TEAMS.USER_ID.eq(userData.getUserId()))
+            .and(USERS_TEAMS.TEAM_ID.eq(teamId))
+            .fetchOne();
+    UsersTeamsRecord newLeaderTeamsRecord =
+        db.selectFrom(USERS_TEAMS)
+            .where(
+                USERS_TEAMS
+                    .USER_ID
+                    .eq(transferOwnershipRequest.getNewLeaderId())
+                    .and(USERS_TEAMS.TEAM_ID.eq(teamId)))
+            .fetchOne();
+    if (oldLeaderTeamsRecord != null
+        && newLeaderTeamsRecord != null
+        && oldLeaderTeamsRecord.getTeamRole() == TeamRole.LEADER
+        && oldLeaderTeamsRecord.getTeamRole() == TeamRole.MEMBER) {
+      newLeaderTeamsRecord.setTeamRole(TeamRole.LEADER);
+      oldLeaderTeamsRecord.setTeamRole(TeamRole.MEMBER);
+      newLeaderTeamsRecord.update();
+      oldLeaderTeamsRecord.update();
+    } else {
+      throw new WrongTeamRoleException(teamId, TeamRole.LEADER);
```

Comment:
We need two different exceptions, one if the calling user isn't a team leader and one if the newLeader isn't a team member

Reply:
I don't think this was addressed. Right now this exception is just saying that the calling user of the route has to be a tean leader, it doesn't say anything about the "newLeaderTeamsRecord" which can also call this to fail so having the one exception for both cases is not correct.